enhancements/installer/openstack-upi.md

```diff
@@ -0,0 +1,325 @@
+title: openstack-upi
+authors:
+  - "@tomassedovic"
+reviewers:
+  - "@mandre"
+  - "@luis5tb"
+  - "@wking"
+  - "@cuppett"
+  - "@iamemilio"
+approvers:
+  - "@sdodson"
+  - "@crawford"
+creation-date: 2019-09-10
+last-updated: 2019-12-10
+status: implementable
+---
+
+# OpenStack UPI
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria
+- [x] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+The initial OpenStack support for OpenShift 4 was around the
+installer-provisioned (IPI) workflow. While this is the most
+convenient approach for the end-users of OpenStack, there are
+situations where it falls short. The goal of IPI is to create the
+infrastructure elements (network, VMs) needed for installing
+OpenShift. In order to reuse pre-existing pieces of infrastructure
+instead, users have to either adapt the [bare metal][baremetal-upi]
+flow or translate concepts from the AWS documentation.
+
+Similar to AWS and GCP, we want to provide documentation and tooling
+for OpenStack specifically -- using what OpenStack users are familiar
+with, and highlighting considerations specific to that platform.
+
+## Motivation
+
+In addition to the general use cases supported by UPI (e.g. control
+over the creation of the resources of the cloud provider, custom
+Ignition configuration), there are additional reasons for having an
+OpenStack-specific UPI.
+
+First, unlike for completely hosted solutions such as AWS or GCP there
+is no single OpenStack.
+
+Some deployments are public clouds akin to AWS while others are only
+available within a single organisation, possibly disconnected from the
+internet. Moreover, the actual services and configuration can vary
+greatly between deployments.
+
+The OpenStack UPI should be able to support scenarios such as these:
+
+* Lack of Swift Object Storage (which we use for serving the bootstrap
+  Ignition config as well as the registry storage)
+* No floating IP addresses (e.g. in deployments with provider
+  networks)
+* Disconnected from the internet
+* Desire to integrate all traffic with a load balancer or DNS that
+  already exists within the organisation running the OpenStack
+
+### Goals
+
+The goal of user-provisioned is to give users more freedom compared to
+installer-provisioned installation.
+
+However, this enhancement is deemed complete once the Installer codebase and
+documentation defines manual steps for achieving the equivalent of a regular
+IPI installation. Further adaptations can and should be added at a later stage,
+on the basis of this work.
+
+* OpenStack UPI documentation available under:
+https://github.com/openshift/installer/blob/master/docs/user/openstack/install_upi.md
+* Ansible playbooks for automating the OpenStack resource creation
+available under:
+https://github.com/openshift/installer/tree/master/upi/openstack
```

Comment:
The contents of installer/upi are not shipped in any artifacts today. You'll notice that th AWS templates are actually copied and pasted into documentation. ex: https://docs.openshift.com/container-platform/4.2/installing/installing_aws_user_infra/installing-aws-user-infra.html#installation-cloudformation-vpc_installing-aws-user-infra

Are we going to be doing the same or does there need to be some new artifact which contains these files?

Reply:
The idea is to ship the playbooks as documentation (having them in the installer repo will come handy for testing them in the CI). That's in our backlog. Should I amend the RFE to mention that intention?

Reply:
Ok, that's fine. I'd just suggest keeping that in mind when developing them as it will be somewhat limiting in how you structure things.